hnn_core/lfp.py

```diff
@@ -0,0 +1,407 @@
+            ax.set_title(title_str)
+
+        plt_show(show)
+        return ax.get_figure()
+
+
+class _LFPArray:
+    """Class for electrode arrays containing NEURON objects
+
+    The handler is set up to maintain a vector of membrane currents at at every
+    inner segment of every section of every cell on each CVODE integration
+    step. In addition, it records a time vector of sample times. This class
+    must be instantiated and attached to the network during the building
+    process. It is used in conjunction with the calculation of extracellular
+    potentials.
+
+    Parameters
+    ----------
+    array : instance of LFPArray
+        Initialised, e.g., by the Network.add_electrode_array()-method
+    cvode : instance of h.CVode
+        Multi order variable time step integration method.
+    """
+    def __init__(self, array, cvode=None):
+        self.positions = array.positions
+        self.n_contacts = len(self.positions)
+        self.sigma = array.sigma
+        self.method = array.method
+
+        # Attach a callback for calculating the potentials at each time step.
+        # Enables fast calculation of transmembrane current (nA) at each
+        # segment. Note that this will run on each rank, so it is safe to use
+        # the extra_scatter_gather-method, which docs say doesn't support
+        # "multiple threads".
+        cvode.extra_scatter_gather(0, self.calc_potential_callback)
+
+    @property
+    def n_samples(self):
+        """Return the length (in samples) of the LFP data."""
+        return int(self._lfp_v.size() / self.n_contacts)
+
+    @property
+    def voltages(self):
+        """The LFP data (n_contacts x n_samples)."""
+        if len(self._lfp_v) > 0:
+            # return as a Neuron Matrix object for efficiency
+            lfpmat = h.Matrix(self.n_contacts, self.n_samples)
+            lfpmat.from_vector(self._lfp_v)
+            return lfpmat
+        else:
+            return None  # simulation not yet run
+
+    @property
+    def times(self):
+        """The sampling time points."""
+        if self._lfp_t.size() > 0:
+            # NB _lfp_t is one sample longer than _lfp_v
+            return self._lfp_t.to_python()[:self.n_samples]
+        else:
+            return None  # simulation not yet run
+
+    def build(self):
+        """Create the Neuron objects needed to record LFPs
+
+        """
+        # ordered list of h.Sections on this rank (if running in parallel)
+        secs_on_rank = _get_sections_on_this_rank()
+        # np.array of number of segments for each section, ordered as above
+        segment_counts = np.array(_get_segment_counts(secs_on_rank))
+
+        # pointers assigned to _ref_i_membrane_ at each EACH internal segment
+        self.imem_ptrvec = h.PtrVector(segment_counts.sum())
+        # placeholder into which pointer values are read on each sim step
+        self.imem_vec_len = int(self.imem_ptrvec.size())
```

Comment:
it's a bit surprising to see these attributes here without seeing them in the init. Can it be initialized to `None` there?

Reply:
I think you could even initialize it with `h.PtrVector`? since this is called when the simulation is being run

Reply:
You're both right. In fact, the `build`-method is totally superfluous, since it's called immediately after instantiation. Removing.